### PR TITLE
Optimizations & simplifications following #25717

### PR DESCRIPTION
**Pull request by sipa**
This contains a list of most-unrelated simplifications and optimizations to the code merged in #25717:
* **Make `ProcessNextHeaders` replace the headers argument**: this allows reusing the same vector storage for input and output headers to `HeadersSync::ProcessNextHeaders`. I find it also natural in the sense that this argument just represents the headers-to-be-processed, both in the caller and the callee, and both before and after the calls.
* **Make `ProcessNextHeaders` support empty headers message**: remove the special case in `ProcessHeadersMessage` dealing with empty headers messages, and instead let `HeadersSync` deal with it correctly.
* **Simplify `TryLowWorkHeadersSync` invocation**: make use of the fact that now `TryLowWorkHeadersSync` is (like `IsContinuationOfLowWorkHeadersSync`) an operation that partially processes headers, it can be invoked in a similar way, bailing out when there is nothing left to do.
* **Avoid an `IsAncestorOfBestHeaderOrTip` call**: just don't call this function when it won't have any effect
* **Compute work from headers without `CBlockIndex`**: avoid the need to construct a `CBlockIndex` object just to compute work for a header, when its `nBits` value suffices for that. Also use some `Span`s where possible.
* **Remove useless `CBlock::GetBlockHeader` (it inherits from it)**: There is no need for a function to convert a `CBlock` to a `CBlockHeader`, as it's a child class of it.

These are not reactions to review feedback, and isn't intended for 24.0.

### CODE REVIEW

**Comment by naumenkogs** [comment]
Concept ACK, light code review ACK e0025e6c21d0429e66d33b0520b71b5180ffc0cc

-----------

I think it would save us some time in the future if you duplicate the descriptions from the PR post in commit messages.

**Review by aureleoules — COMMENTED** [review]
Some additional improvements, feel free to ignore.

This param seem to be unused
 <details>
  <summary>Git diff</summary>

```diff
diff --git a/src/validation.cpp b/src/validation.cpp
index d029b883c..d099f9d90 100644
--- a/src/validation.cpp
+++ b/src/validation.cpp
@@ -3708,7 +3708,7 @@ bool ChainstateManager::ProcessNewBlockHeaders(const std::vector<CBlockHeader>&
     return true;
 }
 
-void ChainstateManager::ReportHeadersPresync(const arith_uint256& work, int64_t height, int64_t timestamp)
+void ChainstateManager::ReportHeadersPresync(int64_t height, int64_t timestamp)
 {
     AssertLockNotHeld(cs_main);
     const auto& chainstate = ActiveChainstate();
diff --git a/src/validation.h b/src/validation.h
index af7e65a54..ac5a3e5c4 100644
--- a/src/validation.h
+++ b/src/validation.h
@@ -1051,7 +1051,7 @@ public:
      *  headers are not yet fed to validation during that time, but validation is (for now)
      *  responsible for logging and signalling through NotifyHeaderTip, so it needs this
      *  information. */
-    void ReportHeadersPresync(const arith_uint256& work, int64_t height, int64_t timestamp);
+    void ReportHeadersPresync(int64_t height, int64_t timestamp);
 
     ~ChainstateManager();
 };
diff --git a/src/net_processing.cpp b/src/net_processing.cpp
index b74675916..1a1c75406 100644
--- a/src/net_processing.cpp
+++ b/src/net_processing.cpp
@@ -4380,7 +4380,7 @@ void PeerManagerImpl::ProcessMessage(CNode& pfrom, const std::string& msg_type,
                 if (it != m_headers_presync_stats.end()) stats = it->second;
             }
             if (stats.second) {
-                m_chainman.ReportHeadersPresync(stats.first, stats.second->first, stats.second->second);
+                m_chainman.ReportHeadersPresync(stats.second->first, stats.second->second);
             }
         }
```
</details> 

Use initializer list instead of constructor body + const
 <details>
  <summary>Git diff</summary>

```diff
diff --git a/src/headerssync.h b/src/headerssync.h
index 566e95ae2..d59df8e0c 100644
--- a/src/headerssync.h
+++ b/src/headerssync.h
@@ -31,16 +31,14 @@ struct CompressedHeader {
         hashMerkleRoot.SetNull();
     }
 
-    CompressedHeader(const CBlockHeader& header)
-    {
-        nVersion = header.nVersion;
-        hashMerkleRoot = header.hashMerkleRoot;
-        nTime = header.nTime;
-        nBits = header.nBits;
-        nNonce = header.nNonce;
-    }
-
-    CBlockHeader GetFullHeader(const uint256& hash_prev_block) {
+    CompressedHeader(const CBlockHeader& header) :
+        nVersion(header.nVersion),
+        hashMerkleRoot(header.hashMerkleRoot),
+        nTime(header.nTime),
+        nBits(header.nBits),
+        nNonce(header.nNonce) {}
+
+    CBlockHeader GetFullHeader(const uint256& hash_prev_block) const {
         CBlockHeader ret;
         ret.nVersion = nVersion;
         ret.hashPrevBlock = hash_prev_block;
```
</details>

File: src/validation.cpp [hunk 1]
```diff
@@ -3432,19 +3432,16 @@ std::vector<unsigned char> ChainstateManager::GenerateCoinbaseCommitment(CBlock&
...
 
-arith_uint256 CalculateHeadersWork(const std::vector<CBlockHeader>& headers)
+arith_uint256 CalculateHeadersWork(Span<const CBlockHeader> headers)
 {
     arith_uint256 total_work{0};
-    for (const CBlockHeader& header : headers) {
-        CBlockIndex dummy(header);
-        total_work += GetBlockProof(dummy);
-    }
+    for (const CBlockHeader& header : headers) total_work += GetBlockProof(header);
```
[the remark this answers is not in the document]
Comment (Sjors) [reply src/validation.cpp:3444]: Without the `total_work +` bit presumably.

File: src/net_processing.cpp [hunk 2]
```diff
@@ -2726,20 +2726,6 @@ void PeerManagerImpl::ProcessHeadersMessage(CNode& pfrom, Peer& peer,
 {
     size_t nCount = headers.size();
```
Comment (ariard) [inline src/net_processing.cpp:2728]: Can be moved just before usage L2774 to save a `size()` invocation.

File: src/net_processing.cpp [hunk 3]
```diff
@@ -2726,20 +2726,6 @@ void PeerManagerImpl::ProcessHeadersMessage(CNode& pfrom, Peer& peer,
...
-    if (nCount == 0) {
-        // Nothing interesting. Stop asking this peers for more headers.
-        // If we were in the middle of headers sync, receiving an empty headers
-        // message suggests that the peer suddenly has nothing to give us
-        // (perhaps it reorged to our chain). Clear download state for this peer.
-        LOCK(peer.m_headers_sync_mutex);
-        if (peer.m_headers_sync) {
-            peer.m_headers_sync.reset(nullptr);
-            LOCK(m_headers_presync_mutex);
-            m_headers_presync_stats.erase(pfrom.GetId());
```
Comment (ariard) [inline src/net_processing.cpp:2744]: Note to reviewers the state is finalized L133 in `ProcessNextHeaders()`.

File: src/net_processing.cpp [hunk 4]
```diff
@@ -2807,14 +2807,14 @@ void PeerManagerImpl::ProcessHeadersMessage(CNode& pfrom, Peer& peer,
...
     // processing.
-    if (!already_validated_work && TryLowWorkHeadersSync(peer, pfrom,
-                chain_start_header, headers)) {
-        // If we successfully started a low-work headers sync, then there
-        // should be no headers to process any further.
-        Assume(headers.empty());
-        return;
+    if (!already_validated_work) {
+        already_validated_work = TryLowWorkHeadersSync(peer, pfrom, chain_start_header, headers);
+        have_headers_sync = already_validated_work;
```
Comment (ariard) [inline src/net_processing.cpp:2812]: Why not assigning directly the result of `TryLowWOrkHeadersSync()` to `have_headers_sync` here? For some code clarity?

**Review by maflcko — COMMENTED** [review]
left some more cleanup ideas

**Comment by fanquake** [comment]
@sipa are you inteterested in adressing the review feedback here?

**Comment by danielabrozzoni** [comment]
I partially grabbed this PR in #32740: I included the last three commits here, and suggestions from comments https://github.com/bitcoin/bitcoin/pull/25968#discussion_r963634928, https://github.com/bitcoin/bitcoin/pull/25968#discussion_r963618705, https://github.com/bitcoin/bitcoin/pull/25968#discussion_r962870412

The first three commits in this PR are still up for grabs:
- Make ProcessNextHeaders replace the headers argument: this allows reusing the same vector storage for input and output headers to HeadersSync::ProcessNextHeaders. I find it also natural in the sense that this argument just represents the headers-to-be-processed, both in the caller and the callee, and both before and after the calls.
- Make ProcessNextHeaders support empty headers message: remove the special case in ProcessHeadersMessage dealing with empty headers messages, and instead let HeadersSync deal with it correctly.
- Simplify TryLowWorkHeadersSync invocation: make use of the fact that now TryLowWorkHeadersSync is (like IsContinuationOfLowWorkHeadersSync) an operation that partially processes headers, it can be invoked in a similar way, bailing out when there is nothing left to do.

**Pull request closed without merging**